### PR TITLE
Kernel 5.6

### PR DESCRIPTION
**Pull request by fepitre**
It keeps the build for 4.0 with the current patch for MSI flag. It handles your upstream patch not yet in 5.6.x and also GCC10 for 4.1.

https://src.fedoraproject.org/rpms/kernel/c/8ced7a7bf5454ba1e020a3ea643848101c7f3b02?branch=master
https://github.com/QubesOS/qubes-issues/issues/5763

### CODE REVIEW

**Comment by xaki23** [comment]
build of 980b309 on fedora-31 for 4.0 failing here:

```
+ /usr/bin/cat /home/user/qubes-src/linux-kernel/0015-xen-pciback-add-attribute-to-allow-MSI-enable-flag-w.patch
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch
1 out of 2 hunks FAILED -- saving rejects to file drivers/xen/xen-pciback/conf_space_capability.c.rej
4 out of 4 hunks FAILED -- saving rejects to file drivers/xen/xen-pciback/pci_stub.c.rej
1 out of 1 hunk FAILED -- saving rejects to file drivers/xen/xen-pciback/pciback.h.rej
error: Bad exit status from /var/tmp/rpm-tmp.2KtsmT (%prep)
```

clean buildvm, using kernel.conf with repo/branch changed.

**Comment by marmarek** [comment]
I have backported support for new attribute name in https://github.com/QubesOS/updates-status/issues/1729, so no need to keep the old patch.

**Comment by fepitre** [comment]
> build of [980b309](https://github.com/QubesOS/qubes-linux-kernel/commit/980b309b0e6145719692a12f54d356fdc988283a) on fedora-31 for 4.0 failing here:
> 
> ```
> + /usr/bin/cat /home/user/qubes-src/linux-kernel/0015-xen-pciback-add-attribute-to-allow-MSI-enable-flag-w.patch
> + /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch
> 1 out of 2 hunks FAILED -- saving rejects to file drivers/xen/xen-pciback/conf_space_capability.c.rej
> 4 out of 4 hunks FAILED -- saving rejects to file drivers/xen/xen-pciback/pci_stub.c.rej
> 1 out of 1 hunk FAILED -- saving rejects to file drivers/xen/xen-pciback/pciback.h.rej
> error: Bad exit status from /var/tmp/rpm-tmp.2KtsmT (%prep)
> ```
> 
> clean buildvm, using kernel.conf with repo/branch changed.

Currently the build is a success under a R4.1 chroot. I'll check for 4.0 but the source are the same.

**Comment by xaki23** [comment]
patch working with 539b147 ... build running.


> I have backported support for new attribute name in [QubesOS/updates-status#1729](https://github.com/QubesOS/updates-status/issues/1729), so no need to keep the old patch.

i am not sure i understand this part. does that xen build/version work with both "old" and "new" kernels? or is this a situation where both need to be new, or both need to be old?

**Comment by marmarek** [comment]
Yes, the linked update works with both old (< 5.6) and new kernel. Before, only old.

File: kernel.spec.in [hunk 1]
```diff
@@ -137,8 +137,11 @@ Patch9: 0010-xen-netfront-add-range-check-for-Tx-response-id.patch
 Patch10: 0011-xen-blkfront-make-local-copy-of-response-before-usin.patch
 Patch11: 0012-xen-blkfront-prepare-request-locally-only-then-put-i.patch
 Patch12: 0013-xen-pcifront-pciback-Update-pciif.h-with-err-and-res.patch
-Patch13: 0014-xen-pciback-add-attribute-to-allow-MSI-enable-flag-w.patch
-Patch14: 0015-xen-events-avoid-NULL-pointer-dereference-in-evtchn-from-irq.patch
+Patch13: 0014-xen-events-avoid-NULL-pointer-dereference-in-evtchn-from-irq.patch
+%if 0%{?fedora} != 25
+Patch14: 0015-xen-pciback-fix-INTERRUPT_TYPE_-defines.patch
```
Comment (marmarek) [inline kernel.spec.in:142]: This patch is needed in all the cases.

File: kernel.spec.in [hunk 2]
```diff
@@ -137,8 +137,11 @@ Patch9: 0010-xen-netfront-add-range-check-for-Tx-response-id.patch
...
+Patch15: 0016-gcc-common.h-params.h-has-been-dropped-in-GCC10.patch
```
Comment (marmarek) [inline kernel.spec.in:143]: Does it break build on R4.0? If not, better avoid release-specific patches.

File: kernel.spec.in [hunk 3]
```diff
@@ -76,6 +76,7 @@ BuildRequires:  gcc >= 5.3.1-6.qubes1
 %else
 %if 0%{?fedora} == 25
 BuildRequires:  gcc >= 6.4.1-1.qubes1
+Conflicts:      xen-libs < 4.8.5-15
```
Comment (marmarek) [inline kernel.spec.in:79]: This needs epoch too (2001)

**Comment by xaki23** [comment]
build of cbbe629 completed by now (>3h wallclock).
installed the updated xen pkgs:
```
[root@dom0 ~]# rpm -qa | grep ^xen
xen-licenses-4.8.5-15.fc25.x86_64
xen-runtime-4.8.5-15.fc25.x86_64
xen-hvm-stubdom-linux-1.0.10-1.fc25.x86_64
xen-hypervisor-4.8.5-15.fc25.x86_64
xen-libs-4.8.5-15.fc25.x86_64
xen-hvm-4.8.5-15.fc25.x86_64
xen-4.8.5-15.fc25.x86_64
```

installing kernel pkg depsfails:
```
[root@dom0 ~]# rpm -i kernel-latest-5.6.3-1.qubes.x86_64.rpm
error: Failed dependencies:
	xen-libs < 2001:4.8.5-15 conflicts with kernel-latest-1000:5.6.3-1.qubes.x86_64
```

not sure why it doesnt like the installed pkg:
```
[root@dom0 ~]# rpm -qi xen-libs
Name        : xen-libs
Epoch       : 2001
Version     : 4.8.5
Release     : 15.fc25
Architecture: x86_64
Install Date: Thu 09 Apr 2020 09:29:37 PM CEST
Group       : Development/Libraries
Size        : 1461616
License     : GPLv2+ and LGPLv2+ and BSD
Signature   : RSA/SHA256, Thu 09 Apr 2020 03:35:00 AM CEST, Key ID 1848792f9e2795e9
Source RPM  : xen-4.8.5-15.fc25.src.rpm
Build Date  : Thu 09 Apr 2020 03:34:19 AM CEST
Build Host  : build-fedora4
...
```

**Comment by xaki23** [comment]
installed it with --nodeps (YOLO!), and it actualy works.
even the gsc/cr50 tpm works (driver is in mainline since 5.5), and the cros EC too (with 5.5.7 ectool just crashed).

so besides the rpm-deps problem, great success. thanks!

going to try building with the qubes/usbip-supported usb-wifi driver tomorrow.

**Comment by fepitre** [comment]
> [root@dom0 ~]# rpm -i kernel-latest-5.6.3-1.qubes.x86_64.rpm
> error: Failed dependencies:
> 	xen-libs < 2001:4.8.5-15 conflicts with kernel-latest-1000:5.6.3-1.qubes.x86_64

It's normally conflicting older versions (see https://github.com/QubesOS/qubes-linux-kernel/pull/203/files#diff-7dd4a560b6384213ecf6ed122ea84abbR79).

**Comment by fepitre** [comment]
@xaki: I should have fixed the conflict with required xen-libs. I also bumped the version to 5.6.4. Do you mind to check it works for you too please?

**Comment by xaki23** [comment]
> @xaki23: I should have fixed the conflict with required xen-libs. I also bumped the version to 5.6.4. Do you mind to check it works for you too please?

heh. i already had a 5.6.4 build running for 1+ hours before the xen-libs commit.
retrying ... 
- build (clean f31 appvm on q4.0) works, 
- installing both dom0 and vm kernel works (== can confirm the deps problem seems fixed), 
- kernel works in appvms (like the one i am typing this in), 
- didnt try the dom0 kernel because i am too lazy to reboot right now. 

overall: +1 for merge

**Pull request merged**